[#4] 로그인 기능 구현

hanwix2 opened the pull request:
- [x] 로그인 구현
 - [x] 잘못된 로그인 입력에 대한 Exception 처리
 - [x] 이미 로그인 되어 있는 경우 처리
 - [x] 회원의 효율적 관리를 위한 User Table 수정 (id와 userId 구분)
 - [x] User Table의 userType 속성 - Enum으로 처리

hanwix2 committed:
[#4] 로그인 기능 구현

- 로그인 기능 구현
- Data Access를 위한 Mapper 인터페이스만 사용 (Repository 삭제)

tlsgkr7416 commented on src/main/java/com/show/showticketingservice/service/UserService.java:
```diff
@@ -1,32 +1,52 @@
     private final Encryptor passwordEncryptor;
 
     public boolean signUp(User user) {
-        if (isIdDuplicated(user.getId()))
+        if (isIdExists(user.getId()))
             return false;
 
         insertUser(user);
         return true;
     }
```
boolean값을 return한다면 메소드명에 is를 붙여주는게 좋을거같아요. 그리고 여기도 boolean을 리턴하는 것보다 예외처리를 해주면 더 좋을거 같네요.

hanwix2 replied:
왜 예외처리가 더 좋은가요?

tlsgkr7416 replied:
예외처리가 더 가독성에 좋을거 같네요!

tlsgkr7416 commented on src/main/resources/mybatis/mappers/UserMapper.xml:
```diff
@@ -3,14 +3,21 @@
 
-    <select id="idDuplicated" parameterType="String" resultType="int">
-        SELECT COUNT(*) FROM User
+    <select id="isIdExists" parameterType="String" resultType="int">
+        SELECT COUNT(*)
+        FROM User
         WHERE id = #{id}
     </select>
```
여기서 결과값이 0 또는 1뿐인데 boolean으로 결과값을 처리해주면 더 좋을거 같네요.

hanwix2 replied:
제가 존재여부를 확인할 때 자주 사용하는 방법인데, 
멘토님께서 취향차이가 있다고 하니 그대로 두겠습니다!

f-lab-dev replied:
그런데 이름이 `isIdExists`이니 저도 boolean이 명확해서 좋아보입니다~

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/UserController.java:
```diff
@@ -15,6 +17,16 @@
+
+        loginService.login(loginRequest);
+
+        return HttpStatusResponse.HTTP_STATUS_OK;
```
단순 200만 보낸다면 리턴타입을 `void`로 해주시면 이 구문은 생략이 가능합니다~

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/SessionLoginService.java:
```diff
@@ -0,0 +1,27 @@
+    private final UserService userService;
+
+    @Override
+    public void login(UserLoginRequest userLoginRequest) {
```
캡슐화까지 깔끔하군요~

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/UserService.java:
```diff
@@ -1,32 +1,52 @@
+    }
+
+    public User getUser(String userIdRequest, String passwordRequest) {
+        if (!isIdExists(userIdRequest)) {
```
이렇게 하면 총 쿼리가 2번 나가게 되는데 mysql에 쿼리를 날리는 비용은 꽤나 비쌉니다~
아래의 `userMapper.getUserByUserId(userIdRequest)` 에서 가져온 결과가 null이면 예외를 던져주는게 낫지 않을까요?

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/UserService.java:
```diff
@@ -1,32 +1,52 @@
-    public boolean isIdDuplicated(String id) {
-        return userRepository.isIdDuplicated(id);
+    public boolean isIdExists(String id) {
+        return userMapper.isIdExists(id) > 0;
```
이거는 취향차이긴 하지만 쿼리 자체에서 boolean을 리턴해줄 수도 있습니다~

f-lab-dev commented on src/main/java/com/show/showticketingservice/tool/advice/AuthenticationExceptionAdvice.java:
```diff
@@ -0,0 +1,31 @@
+
+    @ExceptionHandler(UserIdNotExistsException.class)
+    public ResponseEntity<ExceptionResponse> userIdNotExistsException(final UserIdNotExistsException e, WebRequest request) {
+        log.error(e.getMessage(), e);
```
로거에 `e`를 넘김으로써 `e.getMessage()`는 자동으로 찍힙니다. 즉 중복된 메세지가 보여지게될텐데 다른 메세지를 보여줄 순 없을까요?

f-lab-dev commented on src/main/java/com/show/showticketingservice/tool/encryptor/PasswordEncryptor.java:
```diff
@@ -10,4 +10,9 @@
     public String encrypt(String rawPassword) {
         return BCrypt.hashpw(rawPassword, BCrypt.gensalt());
     }
+
```
이 클래스의 이름은 `PasswordEncryptor`인데 비밀번호의 암호화 방식이 바뀌게 된다면 이 클래스는 그런 확장에 유연한 구조일까요?

hanwix2 committed:
[#4] 로그인 기능 관련 수정

- DB user table 수정: id와 userId 구분 & userType 추가
- user table 수정에 따른 관련 코드(테스트 포함) 수정
- User 모델을 UserRequest와 UserResponse로 구분
- userId 중복 검사에 대한 예외 처리 추가
- 예외처리 중 로거에 보내는 메세지 수정
- UserService의 getUser메소드에서 sql쿼리를 최소화 하기 위한 로직 수정
- PasswordEncryptor 클래스명 BCryctPasswordEncryptor로 변경

hanwix2 committed:
[#4] 중복 로그인 예외처리

이미 로그인 된 상태에서 중복으로 로그인 시도 시 예외처리

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/UserController.java:
```diff
@@ -20,27 +20,21 @@
-        return HttpStatusResponse.HTTP_STATUS_CONFLICT;
+        userService.signUp(userRequest);
+
+        return HttpStatusResponse.CREATED;
```
이거는 `@ResponseStatus` 어노테이션을 붙여줌으로써 생략이 가능합니다

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/SessionLoginService.java:
```diff
@@ -18,9 +18,9 @@
+        UserResponse userResponse = userService.getUser(userLoginRequest.getUserId(), userLoginRequest.getPassword());
 
-        httpSession.setAttribute("userId", user);
+        httpSession.setAttribute("userId", userResponse);
```
`"userId"` 같은 키는 상수로 관리하는게 재사용성 측면에서 더 좋아보입니다~

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/UserService.java:
```diff
@@ -14,39 +16,39 @@
     private final UserMapper userMapper;
 
-    private final Encryptor passwordEncryptor;
+    private final Encryptor bcryptPasswordEncryptor;
```
추후 암호화 방식이 바뀌면 변수명도 바꿔줘야하는걸까요?

hanwix2 replied:
의존성 주입을 필드에서 했을 때 어떤 것을 참조했는지 몰라 명시하려고 했는데
생성자 주입으로 바꿔서 상관 없는 부분이었는데 인지를 못했네요 ㅠ 
수정하겠습니다!

hanwix2 committed:
[#4] UserType 속성 Enum으로 처리 및 피드백 반영

Enumeration 관련
- User의 UserType 속성 Enumeration화
- TypeHandler로 Enum 처리

피드백 관련
- UserConstant.java 생성: session attribute key값 상수 처리
- PasswordEncryptor 클래스, 변수 이름 수정
- @ResponseStatus 어노테이션 사용
- id 중복확인 sql 쿼리문 수정: boolean 형 반환

f-lab-dev commented on src/main/resources/mybatis/mappers/UserMapper.xml:
```diff
@@ -4,19 +4,18 @@
-        SELECT id, userId, password, name, phoneNum, email, address, userType
-        FROM User
+    <select id="getUserByUserId" parameterType="String" resultType="com.show.showticketingservice.model.user.UserResponse" >
+        SELECT *
```
이런 부분은 유지보수를 위해 `*`로 처리하기보다는 어떤 컬럼들을 가져오는지 명시해주는게 좋습니다~

f-lab-dev reviewed (approved):
코멘트 드린 하나만 빼고는 좋습니다~ 👍

hanwix2 committed:
[#4] UserMapper sql 쿼리문 수정

getUserByUserId sql쿼리문 수정 - SELECT 하는 컬럼명 명시
